resolver-dns/src/main/java/io/netty/resolver/dns/DnsQueryIdSpace.java

```diff
@@ -0,0 +1,210 @@
+import io.netty.util.internal.PlatformDependent;
+import io.netty.util.internal.ThreadLocalRandom;
+
+import java.util.Random;
+
+/**
+ * Special data-structure that will allow to retrieve the next query id to use, while still guarantee some sort
+ * of randomness.
+ * The query id will be between 0 (inclusive) and 65535 (inclusive) as defined by the RFC.
+ */
+final class DnsQueryIdSpace {
+    private static final int MAX_ID = 65535;
+    private static final int BUCKETS = 4;
+    // Each bucket is 16kb of size.
+    private static final int BUCKET_SIZE = (MAX_ID + 1) / BUCKETS;
+
+    // If there are other buckets left that have at least 500 usable ids we will drop an unused bucket.
+    private static final int BUCKET_DROP_THRESHOLD = 500;
+    private final DnsQueryIdRange[] idBuckets = new DnsQueryIdRange[BUCKETS];
+
+    DnsQueryIdSpace() {
+        assert idBuckets.length == MathUtil.findNextPositivePowerOfTwo(idBuckets.length);
+        // We start with 1 bucket.
+        idBuckets[0] = newBucket(0);
+    }
+
+    private static DnsQueryIdRange newBucket(int idBucketsIdx) {
+        return new DnsQueryIdRange(BUCKET_SIZE, idBucketsIdx * BUCKET_SIZE);
+    }
+
+    /**
+     * Returns the next ID to use for a query or {@code -1} if there is none left to use.
+     *
+     * @return next id to use.
+     */
+    int nextId() {
+        int freeIdx = -1;
+        for (int bucketIdx = 0; bucketIdx < idBuckets.length; bucketIdx++) {
+            DnsQueryIdRange bucket = idBuckets[bucketIdx];
+            if (bucket != null) {
+                int id = bucket.nextId();
+                if (id != -1) {
+                    return id;
+                }
+            } else if (freeIdx == -1 ||
+                    // Let's make it somehow random which free slot is used.
+                    ThreadLocalRandom.current().nextBoolean()) {
+                    // We have a slot that we can use to create a new bucket if we need to.
+                    freeIdx = bucketIdx;
+            }
+        }
+        if (freeIdx == -1) {
+            // No ids left and no slot left to create a new bucket.
+            return -1;
+        }
+
+        // We still have some slots free to store a new bucket. Let's do this now and use it to generate the next id.
+        DnsQueryIdRange bucket = newBucket(freeIdx);
+        idBuckets[freeIdx] = bucket;
+        int id = bucket.nextId();
+        assert id >= 0;
+        return id;
+    }
+
+    /**
+     * Push back the id, so it can be used again for the next query.
+     *
+     * @param id the id.
+     */
+    void pushId(int id) {
+        int bucketIdx = id / BUCKET_SIZE;
+        if (bucketIdx >= idBuckets.length) {
+            throw new IllegalArgumentException("id too large: " + id);
+        }
+        DnsQueryIdRange bucket = idBuckets[bucketIdx];
+        if (bucket == null) {
+            bucket = newBucket(bucketIdx);
+            idBuckets[bucketIdx] = bucket;
+        }
```

Comment:
Would it be better to assert that the bucket exists? If it doesn't exist how could we have gotten the id in the first place?

Reply:
Yeah maybe you are right...

Reply:
done